Compilation error in latest nightly

There seems to be a compilation error caused by the type resolution in the latest nightly compiler. This broke our CI system https://github.com/rust-ml/linfa/pull/118 yesterday :smile: For `cargo 1.52.0-nightly (1e8703890 2021-03-26)` everything compiles fine, but if I update my toolchain to `cargo 1.53.0-nightly (65d57e6f3 2021-04-04)` I'm getting the following error:
```rust
   Compiling argmin v0.4.1
error[E0282]: type annotations needed
   --> src/solver/particleswarm/mod.rs:129:9
    |
129 | impl<O, P, F> Solver<O> for ParticleSwarm<P, F>
    |         ^ cannot infer type for type parameter `P`

error[E0282]: type annotations needed
  --> src/solver/quasinewton/bfgs.rs:62:12
   |
62 | impl<O, L, H, F> Solver<O> for BFGS<L, H, F>
   |            ^ cannot infer type for type parameter `H`

error[E0282]: type annotations needed
  --> src/solver/quasinewton/dfp.rs:52:12
   |
52 | impl<O, L, H, F> Solver<O> for DFP<L, H, F>
   |            ^ cannot infer type for type parameter `H`

error[E0282]: type annotations needed
  --> src/solver/quasinewton/sr1.rs:78:12
   |
78 | impl<O, L, H, F> Solver<O> for SR1<L, H, F>
   |            ^ cannot infer type for type parameter `H`

error[E0282]: type annotations needed
   --> src/solver/quasinewton/sr1_trustregion.rs:100:9
    |
100 | impl<O, B, R, F> Solver<O> for SR1TrustRegion<B, R, F>
    |         ^ cannot infer type for type parameter `B`
```

--- a/src/solver/particleswarm/mod.rs
+++ b/src/solver/particleswarm/mod.rs
@@ -129,7 +129,7 @@ where
 impl<O, P, F> Solver<O> for ParticleSwarm<P, F>
 where
     O: ArgminOp<Output = F, Param = P, Float = F>,
-    O::Param: Position<F> + DeserializeOwned + Serialize,
+    P: Position<F> + DeserializeOwned + Serialize,
     O::Hessian: Clone + Default,
     F: ArgminFloat,
 {

--- a/src/solver/quasinewton/bfgs.rs
+++ b/src/solver/quasinewton/bfgs.rs
@@ -70,7 +70,7 @@ where
         + ArgminScaledAdd<O::Param, O::Float, O::Param>
         + ArgminNorm<O::Float>
         + ArgminMul<O::Float, O::Param>,
-    O::Hessian: Clone
+    H: Clone
         + Default
         + Debug
         + Serialize

--- a/src/solver/quasinewton/dfp.rs
+++ b/src/solver/quasinewton/dfp.rs
@@ -62,7 +62,7 @@ where
         + ArgminNorm<O::Float>
         + ArgminMul<O::Float, O::Param>
         + ArgminTranspose<O::Param>,
-    O::Hessian: Clone
+    H: Clone
         + Default
         + Serialize
         + DeserializeOwned

--- a/src/solver/quasinewton/sr1.rs
+++ b/src/solver/quasinewton/sr1.rs
@@ -87,7 +87,7 @@ where
         + ArgminDot<O::Param, O::Hessian>
         + ArgminNorm<O::Float>
         + ArgminMul<O::Float, O::Param>,
-    O::Hessian: Debug
+    H: Debug
         + Clone
         + Default
         + Serialize

--- a/src/solver/quasinewton/sr1_trustregion.rs
+++ b/src/solver/quasinewton/sr1_trustregion.rs
@@ -111,7 +111,7 @@ where
         + ArgminNorm<O::Float>
         + ArgminZeroLike
         + ArgminMul<F, O::Param>,
-    O::Hessian: Debug
+    B: Debug
         + Clone
         + Default
         + Serialize

